recipe/meta.yaml

```diff
@@ -13,10 +13,13 @@ source:
   sha256: 06a0f567c1b82746841424d63627e311d93cba4ab0e121f7d9d64701e753467a
 
 build:
-  number: 0
+  number: 1
   string: cuda{{ cuda_compiler_version | replace('.', '') }}py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}  # [cuda_compiler_version != "None"]
   string: cpu_py{{ CONDA_PY }}h{{ PKG_HASH }}_{{ PKG_BUILDNUM }}                                      # [cuda_compiler_version == "None"]
   skip: true  # [win or (linux and cuda_compiler_version in (undefined, 'None', '10.2'))]
+  rpaths:
+    - lib/
+    - {{ SP_DIR }}/torch/lib
```

Comment:
can we try adding:

```
requirements:
   run:
        - cudnn                           # [cuda_compiler_version != "None"]
        - nccl                            # [cuda_compiler_version != "None"]
```

similar to https://github.com/conda-forge/pytorch-cpu-feedstock/blob/main/recipe/meta.yaml#L58 instead?

Reply:
I guess these two lines won't help.

```
2022-06-01T04:06:06.6901147Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libc10.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
2022-06-01T04:06:06.6901798Z .. is this binary repackaging?
2022-06-01T04:06:06.6916172Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch_cpu.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
2022-06-01T04:06:06.6917017Z .. is this binary repackaging?
2022-06-01T04:06:06.6930804Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch_cuda.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
2022-06-01T04:06:06.6931457Z .. is this binary repackaging?
2022-06-01T04:06:06.6945306Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
2022-06-01T04:06:06.6945926Z .. is this binary repackaging?
```
Is a really strange error to me.

Reply:
Truthfully, this is where my knowledge of the build systems ends.

It is unclear to me why we need the rpaths here, and not in torchvision.

https://github.com/conda-forge/torchvision-feedstock/blob/main/recipe/meta.yaml#L27

@conda-forge/cudatoolkit-dev if you could chime in here it owuld be helpful.

Reply:
One reason why I think this is happening is that you are doing alot of custom configuration to find the file `libtoch.so` within the python sitepackages directory. This is somewhat frowned upon and is not OK. this is likely what all these RPATHs are trying to warn you about. I think, that for now, you should do what you need to do to link to this package. to "work". I think adding these "IGNORE RPATHS" is ok.

Reply:
> I guess these two lines won't help.
> 
> ```
> 2022-06-01T04:06:06.6901147Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libc10.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
> 2022-06-01T04:06:06.6901798Z .. is this binary repackaging?
> 2022-06-01T04:06:06.6916172Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch_cpu.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
> 2022-06-01T04:06:06.6917017Z .. is this binary repackaging?
> 2022-06-01T04:06:06.6930804Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch_cuda.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
> 2022-06-01T04:06:06.6931457Z .. is this binary repackaging?
> 2022-06-01T04:06:06.6945306Z WARNING (openmm-torch,lib/libOpenMMTorch.so): $RPATH/libtorch.so not found in packages, sysroot(s) nor the missing_dso_whitelist.
> 2022-06-01T04:06:06.6945926Z .. is this binary repackaging?
> ```
> 
> Is a really strange error to me.

This is not an error, but just warning. Apparently `conda build` ignores the custom rpaths.

Reply:
> It is unclear to me why we need the rpaths here, and not in torchvision.

I guess, the complication is that `libOpenMMTorch.so` is not linked to `libOpenMM.so`, but rather loaded dynamically.

Reply:
> One reason why I think this is happening is that you are doing alot of custom configuration to find the file `libtoch.so` within the python sitepackages directory. This is somewhat frowned upon and is not OK.

PyTorch is detected with CMake module: https://github.com/openmm/openmm-torch/blob/fa102e69f48a80a8fcdfb338f976e8f1ef20182d/CMakeLists.txt#L15. What custom configurations are you referring to?

> this is likely what all these RPATHs are trying to warn you about. I think, that for now, you should do what you need to do to link to this package. to "work". I think adding these "IGNORE RPATHS" is ok.

I think, the root cause is that the PyTorch library needed to use its C++ API is not in `lib` and the loader cannot find them unless a path is specified using RPATH. One could use `LD_LIBRARY_PATH` variable or create a symlinks to the missing libraries, but at the package level RPATH seems to be the most appropriate solution.

Reply:
I think adding the RPATH is correct. My longer term goal is to package `libtorch` to enable this kind of stuff to happen naturally.

The hackery i'm talking about is:
https://github.com/conda-forge/openmm-torch-feedstock/blob/main/recipe/build.sh#L12

![image](https://user-images.githubusercontent.com/90008/172431882-e80e581c-f98b-4476-9aa6-7701686a6a2d.png)

Setting `TORCH_DIR` to this strange directory is the hack.

Though for now, I don't have a better solution. My latest attempts at refactoring the pytorch package are documented in https://github.com/conda-forge/pytorch-cpu-feedstock/issues/108 but there is still a long way to go (especially with GPU) and with the python side of the package too.

Reply:
> My longer term goal is to package libtorch to enable this kind of stuff to happen naturally.

A straightforward way of doing that would be to have it put its libraries in the main `lib` directory instead of a private `lib` directory inside the Python module.

Reply:
Thank you @peastman. Given the constraints of time, I don't have time to do this at the moment. But yes, it is in the longer term goal i outlined in https://github.com/conda-forge/pytorch-cpu-feedstock/issues/108

Reply:
Correct, the only one i'm still unconvinced of is the `lib` RPATH. Not sure where that is coming from. But if you all want to merge, then I don't see "harm" that can be made from this.